### ISSUE TITLE
where I can start for new handwritten data 

### ISSUE THREAD

**Issue by Kiruba-Mediware** [Breta01/handwriting-ocr#53]
I have gone through the entire code they are very efficient and very useful for my need.But I am struggling to Startup, like How can I start for recognizing my handwritten data

**Comment by Breta01** [comment]
If you want to recognize your handwriting, you first have to train machine learning model for it. In this repo I am trying different models. Personally, I would recommend the WordClassifier-CTC. For training this model you need to have training data. For more information about preparing them refer to issue #7

**Comment by Breta01** [comment]
Yes, that's basically it. You will have to change the OCR.ipynb a little bit, because right now it use a different approach. But that will be quite easy.

**Comment by Kiruba-Mediware** [comment]
Now I have created words dataset and it has been trained. @Breta01  Can you please guide me that how to use word recognition approach in OCR.ipynb, Currently it is using character recognition

**Comment by souaissa** [comment]
Nice question. I am wondering the same. Please guide me how to use word recognition approach in OCR.ipynbl?

**Comment by Breta01** [comment]
Ok, here are the changes. First you have to import `extendImg` function from helpers and change the model that is being loaded.
```Python
# Change this line on beginning of the file
from ocr.helpers import implt, resize, extendImg

# Location of the model (I also changed charClass to wordClass)
MODEL_LOC = 'models/word-clas/WordClassifier'
wordClass = Graph(MODEL_LOC)
```
Then you have to change the `Cycler` class which iterates over the bounding boxes of words. Actually, you only have to rewrite the `recognise` function. Also you may have to change few parameters depending on CTC model that you trained.
```Python
def recognise(self, img):
    """ Recognising word and printing it """
    # Specify the image pre-processing that you used for training
    img = imageNorm(
        img,
        60,
        border=False,
        tilt=True,
        hystNorm=True)

    # Set those parameter depending on what you trained
    slider = (60, 2)
    step = 2

    img = cv2.copyMakeBorder(
        img,
        0, 0, slider[1]//2, slider[1]//2,
        cv2.BORDER_CONSTANT,
        value=[0, 0, 0])
    img = extendImg(
        img,
        (img.shape[0], max(-(-img.shape[1] // step) * step, self.slider[1] + step)))
    length = (img.shape[1]-slider[1]) // step
    input_seq = np.zeros((1, length, slider[0] * slider[1]), dtype=np.float32)
    input_seq[0][:] = [img[:, loc*step: loc*step + slider[1]].flatten()
                       for loc in range(length)]
    input_seq = input_seq.swapaxes(0, 1)
            
    pred = wordClass.eval_feed({
        'inputs:0': input_seq,
        'inputs_length:0': [length],
        'keep_prob:0': 1})[0]
            
    word = ''
    for i in pred:
        if word == 0 and i != 0:
            break
        else:
            word += idx2char(i)
    print("Word: " + word)
    return word
```
For more details you can look on implementation of [`OCR-Evaluator.ipynb`](https://github.com/Breta01/handwriting-ocr/blob/master/OCR-Evaluator.ipynb)

**Comment by Breta01** [comment]
OK, so now you have 37 images of words (or something else?). In case of words, the best approach would be to tune parameters of already existing model. Try following file (it may require some changes). 
[WordClassifier-CTC-Retrain.zip](https://github.com/Breta01/handwriting-ocr/files/2429241/WordClassifier-CTC-Retrain.zip)

**Comment by Kiruba-Mediware** [comment]
@Breta01 , I have gone through your file , 
trainImages, trainLabels = loadWordsData(
    ['data/words_arch/dataset/train/'],
    loadGaplines=False)
testImages, testLabels = loadWordsData(
    ['data/words_arch/dataset/train/eval/'],
    loadGaplines=False)
I don't have above kind of data set, I am having just a word dataset that is created through DM file.
Also In your file I don't see anywhere model is saved, it is using the existing model.
In my case I need to train a dataset having 37 words .
Can you give some more clarity on that file u sent me "WordClassifier-CTC-Retrain.zip"

**Comment by Breta01** [comment]
Ok, my mistake, just remove everything which is after training cell. Alse `words_arch` was just name for one of my words datasets. You can raplace it with location of your words. With saving you will have to edit `Graph` class from `tfhelpers.py` and then run `wordClass.save(your_location)` after training:
```Python
class Graph():
    """ Loading and running isolated tf graph """
    def __init__(self, loc, operation='activation', input_name='x'):
        """
        loc: location of file containing saved model
        operation: name of operation for running the model
        input_name: name of input placeholder
        """
        self.input = input_name + ":0"
        self.graph = tf.Graph()
        self.sess = tf.Session(graph=self.graph)
        with self.graph.as_default():
            self.saver = tf.train.import_meta_graph(loc + '.meta', clear_devices=True)
            self.saver.restore(self.sess, loc)
            self.op = self.graph.get_operation_by_name(operation).outputs[0]

    def save(self, location):
        """ Save current model """
        self.saver.save(self.sess, location)
```

I didn't test the above code. There might be some issues.
